cvmfs/catalog_mgr_client.cc

```diff
@@ -0,0 +1,262 @@
+/**
+ * This file is part of the CernVM file system.
+ */
+
+#include "cvmfs_config.h"
+#include "catalog_mgr_client.h"
+
+#include "cache.h"
+#include "download.h"
+#include "fetch.h"
+#include "manifest.h"
+#include "quota.h"
+#include "signature.h"
+#include "statistics.h"
+
+using namespace std;  // NOLINT
+
+namespace catalog {
+
+/**
+ * Triggered when the catalog is attached (db file opened)
+ */
+void ClientCatalogManager::ActivateCatalog(Catalog *catalog) {
+  const Counters &counters = const_cast<const Catalog*>(catalog)->GetCounters();
+  if (catalog->IsRoot()) {
+    all_inodes_ = counters.GetAllEntries();
+  }
+  loaded_inodes_ += counters.GetSelfEntries();
+}
+
+
+ClientCatalogManager::ClientCatalogManager(
+  const string &repo_name,
+  cvmfs::Fetcher *fetcher,
+  signature::SignatureManager *signature_mgr,
+  perf::Statistics *statistics)
+  : AbstractCatalogManager(statistics)
+  , repo_name_(repo_name)
+  , fetcher_(fetcher)
+  , signature_mgr_(signature_mgr)
+  , offline_mode_(false)
+  , all_inodes_(0)
+  , loaded_inodes_(0)
+{
+  LogCvmfs(kLogCatalog, kLogDebug, "constructing client catalog manager");
+  n_certificate_hits_ = statistics->Register("cache.n_certificate_hits",
+    "Number of certificate hits");
+  n_certificate_misses_ = statistics->Register("cache.n_certificate_misses",
+    "Number of certificate misses");
+}
+
+
+ClientCatalogManager::~ClientCatalogManager() {
+  LogCvmfs(kLogCache, kLogDebug, "unpinning / unloading all catalogs");
+
+  for (map<PathString, shash::Any>::iterator i = mounted_catalogs_.begin(),
+       iend = mounted_catalogs_.end(); i != iend; ++i)
+  {
+    fetcher_->cache_mgr()->quota_mgr()->Unpin(i->second);
+  }
+}
+
+
+Catalog *ClientCatalogManager::CreateCatalog(
+  const PathString  &mountpoint,
+  const shash::Any  &catalog_hash,
+  catalog::Catalog  *parent_catalog
+) {
+  mounted_catalogs_[mountpoint] = loaded_catalogs_[mountpoint];
+  loaded_catalogs_.erase(mountpoint);
+  return new Catalog(mountpoint, catalog_hash, parent_catalog);
+}
+
+
+shash::Any ClientCatalogManager::GetRootHash() {
+  ReadLock();
+  shash::Any result = mounted_catalogs_[PathString("", 0)];
+  Unlock();
+  return result;
+}
+
+
+/**
+ * Specialized initialization that uses a fixed root hash.
+ */
+bool ClientCatalogManager::InitFixed(const shash::Any &root_hash) {
+  LogCvmfs(kLogCatalog, kLogDebug, "Initialize catalog with root hash %s",
+           root_hash.ToString().c_str());
+  WriteLock();
+  bool attached = MountCatalog(PathString("", 0), root_hash, NULL);
+  Unlock();
+
+  if (!attached) {
+    LogCvmfs(kLogCatalog, kLogDebug, "failed to initialize root catalog");
+  }
+
+  return attached;
+}
+
+
+LoadError ClientCatalogManager::LoadCatalog(
```

Comment:
I would call it 'loose coupling' or 'separation of concerns'. In fact the fetching is not so much different on the server, only that we have no cache there. Download and cleanup could very well be shared, since the server uses the `DownloadManager` already.

Right now the `ClientCatalogManager` is connected to pretty much everything, since it knows about (at least some of) the internals of `CacheManager`, `PosixCacheManager`, `QuotaManager`, `Fetcher`, `Catalog`, `Manifest`, `manifest::Fetch` and `SignatureManager`. It even knows about the `.cvmfschecksum` file that resides in certain cache directories and that CVMFS usually switches to the cache directory at startup - admittedly, this is marked as a *TODO hack*. Pretty many concerns and lots of knowledge for a single class - or even a single method - if you ask me.

Please don't get me wrong: I just wanted to point that out; I agree that changing all that would be another major effort. The changes in this Pull Request are certainly a step in the right direction.